src/AudacityMirProject.cpp

```diff
@@ -0,0 +1,119 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  AudacityMirProject.cpp
+
+  Matthieu Hodgkinson
+
+**********************************************************************/
+#include "AudacityMirProject.h"
+#include "AdornedRulerPanel.h"
+#include "AudacityDontAskAgainMessageDialog.h"
+#include "Project.h"
+#include "ProjectHistory.h"
+#include "ProjectTimeSignature.h"
+#include "ProjectWindows.h"
+#include "TimeDisplayMode.h"
+#include "WaveTrack.h"
+#include "prefs/ImportExportPrefs.h"
+
+namespace
+{
+auto FormatTempo(double value)
+{
+   auto rounded = std::to_string(std::round(value * 100) / 100);
+   rounded.erase(rounded.find_last_not_of('0') + 1, std::string::npos);
+   rounded.erase(rounded.find_last_not_of('.') + 1, std::string::npos);
+   return rounded;
+}
+} // namespace
+
+AudacityMirProject::AudacityMirProject(AudacityProject& project)
+    : mProject { project }
+    , mImportedOnEmptyProject { TrackList::Get(mProject)
+                                   .Any<WaveTrack>()
+                                   .empty() }
+    , mProjectTempo { ProjectTimeSignature::Get(mProject).GetTempo() }
+{
+}
+
+AudacityMirProject::~AudacityMirProject()
+{
+   auto& ph = ProjectHistory::Get(mProject);
```

Comment:
Is it really safe? I see that there could be `UserException` thrown at the same scope where `AudacityMirProject` is created

Reply:
What is it that could throw?

Reply:
`reportProgress` in `RunTempoDetection`

Reply:
Ok, I see the exception.
Should be fine: if the user cancels the analysis, then there won't be time to modify the project, and the ctor should return early.